src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -511,14 +521,168 @@ private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<Symbol
+        /// Use TupleNamesCheckKind.Shallow when the caller is already visiting constituent types.
+        /// </summary>
+        private static bool CheckTupleNamesConstraints(TypeSymbol type, ArrayBuilder<TypeParameterDiagnosticInfo> builder, TupleNamesCheckKind tupleNamesCheckKind)
+        {
+            // The skip list is used to avoid duplicate checks. It also avoid circular checks, as in `Derived : Base<Derived>`.
+            var skipList = PooledHashSet<TypeSymbol>.GetInstance();
+            bool result = CheckTupleNamesConstraints(type, builder, tupleNamesCheckKind, skipList);
+            skipList.Free();
+            return result;
+        }
+
+        private static bool CheckTupleNamesConstraints(TypeSymbol type, ArrayBuilder<TypeParameterDiagnosticInfo> builder, TupleNamesCheckKind tupleNamesCheckKind, HashSet<TypeSymbol> skipList)
+        {
+            // Note this only checks for tuple name errors introduced by referencing this type.
+            //
+            // So if the type does not include any tuples, its reference cannot introduce an error.
+            // But its declaration could. For instance:
+            //     `class Error : Base, I<(int notA, int notB)>`
+            //
+            // Similarly, if the base type doesn't use type arguments from the type, then the reference of this type cannot introduce an error.
+            // But the base type may already have one (reported elsewhere). For instance:
+            //     `class C : BaseCanBeSkipped<int>, ITest`
+            //     `class D : BaseNeedsChecking<T>, ITest`
+
+            Debug.Assert((object)type != null && (object)skipList != null);
+            if (skipList.Contains(type) || !type.ContainsTuple())
+            {
+                return true;
+            }
+
+            switch (tupleNamesCheckKind)
+            {
+                case TupleNamesCheckKind.Shallow:
+                    {
+                        skipList.Add(type);
+                        if (!CheckTupleNamesConstraintsCore(type, builder))
+                        {
+                            return false;
+                        }
+
+                        // Walk up the base types and check each inheritence level.
+                        // We can cut short if the original definition's base doesn't contain type parameters.
+                        TypeSymbol currentType = type;
+                        NamedTypeSymbol baseType = currentType.BaseTypeNoUseSiteDiagnostics;
+                        bool result = true;
+                        while ((object)baseType != null && currentType.OriginalDefinition.BaseTypeNoUseSiteDiagnostics.ContainsTypeParameter())
+                        {
+                            result &= CheckTupleNamesConstraints(baseType, builder, TupleNamesCheckKind.Deep, skipList);
+                            currentType = baseType;
+                            baseType = currentType.BaseTypeNoUseSiteDiagnostics;
```

Comment:
Actually it doesn't look like implementation of TupleNamesCheckKind.Deep check has to be modified at all, it is sufficient to get rid of the loop here.

---
In reply to: [93530572](https://github.com/dotnet/roslyn/pull/15809#discussion_r93530572) [](ancestors = 93530572)